<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <link rel="stylesheet" href="http://www.ncbi.nlm.nih.gov/corehtml/ncbi2.css" type="text/css" />
  <style type="text/css">
  .blast h3 { font-size: 14px; }
  .blast a { text-decoration: none; }
  .blast li {  font-size: 12px; }
  .fix-width { font-family: courier; font-size: 80%; }
  .pink { font-weight: bold; font-size: 100%; color: #ff8888; font-family: arial, helvetica, sans-serif; }
  .text2 {font-size: 10pt;font-family: arial,helvetica,sans-serif;}
  .tbl_title{font-weight: bold; font-size: 10pt; color: #0000FF; font-family: arial, helvetica, sans-serif;}
  </style>
<title>blast client</title>
</head>

<body>
<table bgcolor="#eeeeff" width="600" class="text2">
<tbody>
<tr><td align="center" class="tbl_title">Program Option for Netblast (blastcl3)</td></tr>
<tr><td align="center">Tao Tao, Ph.D.<br />User Service<br />NCBI, NLM, NIH</td></tr>

<tr><td class="tbl_title">TOC</td></tr>
<tr><td><ul>
	<li><a href="#1">1. Introduction</a></li>
	<li><a href="#2">2. Installation and setup</a></li>
	<li><a href="#3">3. Firewall settings</a></li>
	<li><a href="#4">4. Options and their accepted values</a></li>
	<li><a href="#5">5. Practical usage examples</a>
		<ul>
			<li><a href="#5.1">5.1 General nucleotide searches</a></li>
			<li><a href="#5.2">5.2 General protein searches</a></li>
			<li><a href="#5.3">5.3 Translated BLAST searches</a>
				<ul>
					<li><a href="#5.3.1">5.3.1 blastx</a></li>
					<li><a href="#5.3.2">5.3.2 tblastn</a></li>
					<li><a href="#5.3.3">5.3.3 tblastx</a></li>
				</ul></li>
			<li><a href="#5.4">5.4 Genome BLAST searches</a>
				<ul>
					<li><a href="#5.4.1">5.4.1 Microbial Genomes and Other Eukaryote Genomes</a></li>
					<li><a href="#5.4.2">5.4.2 Higher Genomes</a></li>
				</ul></li>
		</ul></li>
	<li><a href="#6">6. Trouble shooting and technical assistance</a>
		<ul>
			<li><a href="#6.1">6.1 Errors and warnings</a></li>
			<li><a href="#6.2">6.2 Technical assistance</a></li>
		</ul></li>
</td></tr>
<tr><td>&nbsp;</td></tr>

<tr><td class="tbl_title"><a name="1">1. Introduction</a></td></tr>
<tr><td>
NCBI BLAST web server provides a convenient and user friendly way for individuals to search
their queries against different public sequence databases. This server, however, does have some
limitation. For example, one will not be able to perform large scale batch searches from most of
the BLAST pages and the program selection for some of the available databases is limited.
BLAST client provides a way to circumvent those limitations.
<p/>
The client bypasses web browser and interacts directly with the NCBI BLAST server that powers
the NCBI web BLAST service (<a href="http://www.ncbi.nlm.nih.gov/BLAST/">www.ncbi.nlm.nih.gov/BLAST/</a>). It performs the batch search with
multiple sequences by taking one query sequence from the input file (with multiple FASTA
formatted sequences), formulating the search according to the command line, and sending
the search through the internet connection to NCBI BLAST server for processing.  The program
receives the search result from blast server and saves it to a local file specified by the command
line. The program loops through all the queries in the input file till all are searched.
<p/>
This program has no graphic user interface (GUI) and must be executed from command line under a
terminal window. Users control the program through command line options. Detailed list of
command line options are in <a href="#4">Section 4</a>. For usages and situation examples, see 
<a href="5">Section 5</a>.
<p/>
</td></tr>

<tr><td class="tbl_title"><a name="2">2. Installation and setup</a></td></tr>
<tr><td>
NCBI provides BLAST client as an archive separate from that of the standalone (blast initialed)
or server blast (wwwblast initialed) package. This archive is available for common platforms as
netblast initialed files. They can be found at:

<blockquote>
<a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/">
ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/</a>
</blockquote>

For Linux or Unix environment, installation is straight forward. One can place the archive in a
desired directory and extract the archive using the following command line:

<blockquote>tar zxvf netblast-##-**.tar.gz</blockquote>

The resulted netblast-## directory contains bin, doc, and data subdirectories. The program,
blastcl3, is under the bin subdirectory. The matrices BLAST needs for protein alignments are
under the data subdirectory, while the doc subdirectory contains netblast.html (this file) and
firewall.html with more information on configuration under firewall settings.
<p/>
The package for Windows can be extracted using WinZip. It does not have this directory structure.
<p/>
</td></tr>
<tr><td>&nbsp;</td></tr>

<tr><td class="tbl_title"><a name="3">3. Firewall settings</a></td></tr>
<tr><td>
The setup for NCBI network clients has been greatly simplified. If you are not behind a firewall no
further action is required. If you are behind a firewall, and already use Sequin or Entrez, or if your
system administrator has already performed the setup, then you should be able to start
performing searches immediately after installation.  Otherwise, your will need to make sure that
the following IP address/port combinations are open in the firewall configuration.
<p/>

<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">
	Table 3. Firewall Ports Needed by BLAST Client for NCBI Connection</td></tr>
<tr><td><b>IP Address</b></td><td><b>Port Number</b></td></tr>
<tr><td>130.14.29.112 </td><td>5861 </td></tr>
<tr><td>130.14.29.112 </td><td>5862 </td></tr>
<tr><td>130.14.29.112 </td><td>5863 </td></tr>
<tr><td colspan="2" class="medium2">Note Please refer to 'firewall.html' included in the package 
	for details.</td></tr>
</table>
<p/>
In addition to this, you also need to create an .ncbirc file placed in the home directory to instruct
blastcl3 how to make the connection to NCBI. For PC running Windows, the file is named ncbi.ini
which should be placed under the windows directory. A sample .ncbirc file is provided in the text
box below for your reference.

<blockquote>
[NCBI]<br/>
DATA=/home/johndoe/netblast-2.2.12/data<br/>
<br/>
[CONN]<br/>
FIREWALL=TRUE<br/>
<br/>
[NET_SERV]<br/>
SRV_CONN_MODE=SERVICE
</blockquote>

As an alternative to blastcl3, NCBI BLAST web server also supports URL API, which uses URL
encoded command to interact with Blast.cgi directly to "Put" search requests or to "Get" search
results. For details on the standard commands, please refer to the online document at:
<blockquote>
<a href="http://www.ncbi.nlm.nih.gov/blast/Doc/urlapi.html">
www.ncbi.nlm.nih.gov/blast/Doc/urlapi.html</a></blockquote>
</td></tr>


<tr><td>&nbsp;</td></tr>
<tr><td class="tbl_title"><a name="4">4. Options and their accepted values</a></td></tr>
<tr><td>
As mentioned before in Section 1, blastcl3 has no GUI and works only under a command
terminal. Users execute the the program by issuing command lines, and controls the way blast
search is done through options in the command line. The command line options for this program
are listed here individually below. The options commonly adjusted during actual searches are: -i, -
d, -p, -o, -e, -F, -u, -b, -v, -m, and -n. The first four are mandatory.

<p/>
<table width="600" border="1" class="text">
<tr><td colspan="4" class="tbl_title">Table 4.1</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink" colspan="3">-p </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="3">Specifies which program to run </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="3">None, mandatory </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="3">String </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="3">To run blastn program use: -p blastn </td></tr>
<tr><td class="tbl_title" rowspan="7">Note</td><td colspan="3" class="tbl_title">Program string options and type of search they specify </td></tr>
<tr><td>Program</td><td>Query</td><td>DB</td></tr>
<tr><td>blastn</td><td>nucleotide</td><td>nucleotide</td></tr>
<tr><td>blastp</td><td>protein</td><td>protein</td></tr>
<tr><td>blastx</td><td>nucleotide, translated</td><td>Protein</td></tr>
<tr><td>tblastn</td><td>protein</td><td>nucleotide, translated</td></tr>
<tr><td>tblastx</td><td>nucleotide, translated</td><td>nucleotide, translated</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.2</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-d </td></tr>
<tr><td class="tbl_title">Function</td><td>Specifies database(s) to be searched</td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">nr </td></tr>
<tr><td class="tbl_title">Input Format</td><td>String </td></tr>
<tr><td class="tbl_title">Example</td><td>One can search against multiple databases can be specified in command
line. To search nr and est at together use: -d "nr est" </td></tr>
<tr><td class="tbl_title">Note</td><td>Be conservative. Search against large databases may not complete due
to CPU time limit, which is set at one hour. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.3</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-i </td></tr>
<tr><td class="tbl_title">Function</td><td>Specifies input query file </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">stdin </td></tr>
<tr><td class="tbl_title">Input Format</td><td>String, mandatory </td></tr>
<tr><td class="tbl_title">Example</td><td>To use sequences from query.txt as query, use -i query.txt </td></tr>
<tr><td class="tbl_title">Note</td><td>One should use the complete file name WITH its extension.
To use stdin default, omit the -i and redirect using: &lt; mito.txt</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.4</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-e  </td></tr>
<tr><td class="tbl_title">Function</td><td>Specifies Expect value cutoff </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">10 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>Real </td></tr>
<tr><td class="tbl_title">Example</td><td>To make the search more stringent, one can use: -e 0.001 </td></tr>
<tr><td class="tbl_title">Note</td><td>Accepted formats are integer, fraction, decimal, exponential and scientific
notation. To set the cutoff to 2&#215;10-20, use -e 2e-20 </td></tr>
</table>

<p/>
<table width="600" border="1" class="text">
<tr><td colspan="3" class="tbl_title">Table 4.5</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td colspan="2" class="pink">-m </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="2">Specifies alignment view option </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="2">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="2">Integer </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="2">To display the result in XML form use: -m 7 </td></tr>
<tr><td class="tbl_title" rowspan="13">Note</td><td colspan="2" class="tbl_title">Option values and the output formats they specify </td></tr>
<tr><td>0</td><td>Pairwise</td></tr>
<tr><td>1</td><td>query-anchored showing identities</td></tr>
<tr><td>2</td><td>query-anchored no identities</td></tr>
<tr><td>3</td><td>flat query-anchored, show identities</td></tr>
<tr><td>4</td><td>flat query-anchored, no identities</td></tr>
<tr><td>5</td><td>query-anchored no identities and blunt ends</td></tr>
<tr><td>6</td><td>flat query-anchored, no identities and blunt ends</td></tr>
<tr><td>7</td><td>XML Blast output</td></tr>
<tr><td>8</td><td>tabular (not post processing)</td></tr>
<tr><td>9</td><td>tabular with comment lines (post-processed, sorted)</td></tr>
<tr><td>10</td><td>ASN, text</td></tr>
<tr><td>11</td><td>ASN, binary</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.6</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-o </td></tr>
<tr><td class="tbl_title">Function</td><td>Specifies result output file </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">stdout (print to screen) </td></tr>
<tr><td class="tbl_title">Input Format</td><td>String [file name] </td></tr>
<tr><td class="tbl_title">Example</td><td>To save result in out.txt use: -o out.txt </td></tr>
<tr><td class="tbl_title">Note</td><td>-p, -i, -d, -o are the core options needed for a blastcl3 search. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.7</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-F </td></tr>
<tr><td class="tbl_title">Function</td><td>Specifies which filter(s) to use to mask query sequence </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">T (DUST for nucleotide, SEG for protein) </td></tr>
<tr><td class="tbl_title">Input Format</td><td>String </td></tr>
<tr><td class="tbl_title">Example</td><td>To filter low complexity and lookup table only, use: -F "m L" </td></tr>
<tr><td class="tbl_title">Note</td><td>Accepted strings: T, F, D, L, R, V, S, C, and m.<br/>
m in -F stands for masking for lookup table only, which enables blast to<br/>
display the masked region in the alignment. L stands for Low complexity,<br/>
D stands for DUST. R stands for human Repeats, V stands for Vector.<br/>
<br/>
S stands for SEG, which has other user specifiable values:<br/>
-F "S 10 1.0 1.5" SEG filter: window=10; low cut=1; high cut=1.5.<br/>
C stands for COIL, which also has user specifiable values:<br/>
-F "C 28 40 32" COIL filter: window=22; cutoff=40; linker=32.<br/>
<br/>
To run SEG and COIL filter together, use: -F "S; C"<br/>
To mask lookup table only, add m: -F m "S; C"<br/>
<br/>
To mask repeat sequences use: -F R or -F "m R"<br/>
To combine all together, use: -F "m L;R"<br/>
To mask vector filter, use: -F V<br/>
To call rodent repeat filter, use: -F "R -d rodent.lib"</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.8</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-G </td></tr>
<tr><td class="tbl_title">Function</td><td>Cost to open a gap </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To increase the gap open penalty to 10, use: -G 10 </td></tr>
<tr><td class="tbl_title">Note</td><td>Zero invokes default (5) for blastn. It varies for blastp, blastx, tblastn, and
tblastx. In protein searches, only a controlled set of -G/-E value pairs are acceptable for a
given scoring matrix. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.9</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-E </td></tr>
<tr><td class="tbl_title">Function</td><td>Cost to extend a gap </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To increase the gap extension penalty to 4, use: -E 4 </td></tr>
<tr><td class="tbl_title">Note</td><td>Zero invokes default or 2 for blastn. Varies for blastp, blastx, tblastn, and
tblastx. In protein searches, only a controlled set of -G/-E value pairs are acceptable for a
given scoring matrix. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td class="tbl_title" colspan="6">Table 4.10</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td colspan="5" class="pink">-X </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="5">X dropoff value for gapped alignment (in bits)</td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="5">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="5">[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="5">To increase the gapped alignment dropoff to 40, use: -X 40 </td></tr>
<tr><td class="tbl_title" rowspan="3">Note</td><td colspan="5">Gapped Alignment Dropoff Default Setting  (in bits) </td></tr>
<tr><td>Program</td><td>blastn</td><td>megablast</td><td>tblastx</td><td>others</td></tr>
<tr><td>Value</td><td>30</td><td>20</td><td>0</td><td>15</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.11</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-I (capital i)</td></tr>
<tr><td class="tbl_title">Function</td><td>Show GI in definition line </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">F </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[T/F] </td></tr>
<tr><td class="tbl_title">Example</td><td>To activate the GI display use: -I T </td></tr>
<tr><td class="tbl_title">Note</td><td>Sample display:<br/>T: gi|223046|prf||0410468A... <br/>F: prf||0410468A... </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.12</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-q </td></tr>
<tr><td class="tbl_title">Function</td><td>Penalty for a nucleotide mismatch </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">-3 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To set penalty to -2, use: -q -2 </td></tr>
<tr><td class="tbl_title">Note</td><td>For blastn only, different -r/-q ratios are optimal for aligning sequences
with different percentage of similarities. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.13</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-r </td></tr>
<tr><td class="tbl_title">Function</td><td>Reward for a nucleotide match </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">1 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To increase the reward to 2, use: -r 2 </td></tr>
<tr><td class="tbl_title">Note</td><td>For blastn only. Others use external scoring matrix to determine this. See
-M table in blastall for more details. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.14</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-v </td></tr>
<tr><td class="tbl_title">Function</td><td>Number of database sequences to show one-line descriptions for </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">500 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To increase the descriptions displayed to 1000 use: -v 1000 </td></tr>
<tr><td class="tbl_title">Note</td><td>Web counterpart is "Descriptions" </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.15</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-b </td></tr>
<tr><td class="tbl_title">Function</td><td>Number of sequences with alignments to show </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">[Integer] </td></tr>
<tr><td class="tbl_title">Input Format</td><td>250 </td></tr>
<tr><td class="tbl_title">Example</td><td>To increase the alignment displayed to 1000 use: -b 1000 </td></tr>
<tr><td class="tbl_title">Note</td><td>Upper limit is 200000. Web counterpart: "Alignments".
This is NOT the total number of alignment segments or high scoring pairs
(HSPs). Rather it is the number of database sequences with HSP(s) to
the query. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="8" class="tbl_title">Table 4.16</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink" colspan="7">-f </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="7">Threshold for extending hits </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="7">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="7">Integer </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="7">To increase this threshold to 15, use: -f 15 </td></tr>
<tr><td class="tbl_title" rowspan="3">Note</td><td colspan="7">Default if set to zero, not used by blastn or megablast. Extension Threshold Default Settings </td></tr>
<tr><td>Program</td><td>blastp</td><td>blasn</td><td>blastx</td><td>tblastn</td><td>tblastx</td><td>megablast</td></tr>
<tr><td>Value</td><td>11</td><td>0</td><td>12</td><td>13</td><td>13</td><td>0</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.17</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-g </td></tr>
<tr><td class="tbl_title">Function</td><td>Perform gapped alignment</td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">T </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[T/F] </td></tr>
<tr><td class="tbl_title">Example</td><td>To do only ungapped alignment, use: -g F </td></tr>
<tr><td class="tbl_title">Note</td><td>Default is gapped alignment, not available with tblastx. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.18</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-Q </td></tr>
<tr><td class="tbl_title">Function</td><td>Query genetic code to use </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">1 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To set the genetic code (translation table) to 14, use: -Q 14 </td></tr>
<tr><td class="tbl_title">Note</td><td>This determines which translation table to use on query in translated
blastx and tblastx searches. Default is universal codon. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.19</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-D </td></tr>
<tr><td class="tbl_title">Function</td><td>DB Genetic code </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">1 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To set the genetic code (translation table) to 14, use: -D 14 </td></tr>
<tr><td class="tbl_title">Note</td><td>Determines which translation table to use for the database in tblastn and
tblastx search. See details at: <a href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.20</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-a </td></tr>
<tr><td class="tbl_title">Function</td><td>Number of processors to use</td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">1 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To change this to two CPUs, use: -a 2 </td></tr>
<tr><td class="tbl_title">Note</td><td>From 1 up to the number of CPUs available. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.21</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-O </td></tr>
<tr><td class="tbl_title">Function</td><td>To save SeqAlign object </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">N/A </td></tr>
<tr><td class="tbl_title">Input Format</td><td>String [File Out] </td></tr>
<tr><td class="tbl_title">Example</td><td>To save SeqAlign object to blast_seqalign, use: -O blast_seqalign </td></tr>
<tr><td class="tbl_title">Note</td><td>User can use the output to reformat the result into different format using
NCBI toolkit function. See <a href="ftp://ftp.ncbi.nih.gov/blast/demo/">/blast/demo/</a> subdirectory
for more information. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.22</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-J </td></tr>
<tr><td class="tbl_title">Function</td><td>Believe the query definition line </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">F </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[T/F] </td></tr>
<tr><td class="tbl_title">Example</td><td>To set this to true, use: -J T </td></tr>
<tr><td class="tbl_title">Note</td><td>Default set to false since query file definition lines may not follow NCBI
convention. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.23</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-M </td></tr>
<tr><td class="tbl_title">Function</td><td>Protein scoring matrix to use </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">BLOSUM62 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[String] </td></tr>
<tr><td class="tbl_title">Example</td><td>To change this to PAM30, use: -M PAM30 </td></tr>
<tr><td class="tbl_title">Note</td><td>Accepted value: BLOSUM45, BLOSUM62, BLOSUM80, PAM30, or PAM70. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td class="tbl_title" colspan="5">Table 4.24</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink" colspan="4">-W </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="4">Word size </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="4">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="4">[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="4">To set word size to 32, use: -W 32 </td></tr>
<tr><td class="tbl_title" rowspan="3">Note</td><td colspan="4">Word size setting for different programs </td></tr>
<tr><td>Program</td><td>blastn</td><td>megablast</td><td>all others</td></tr>
<tr><td>Value</td><td>11</td><td>28</td><td>3</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.25</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-z </td></tr>
<tr><td class="tbl_title">Function</td><td>Effective length of the database </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Real] </td></tr>
<tr><td class="tbl_title">Example</td><td>To set this to 10000000, use: -z 10000000 </td></tr>
<tr><td class="tbl_title">Note</td><td>Use zero for the actual database size. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.26</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-K </td></tr>
<tr><td class="tbl_title">Function</td><td>Number of best hits from a region to keep</td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td> To keep 200 hits, use: -K 200</td></tr>
<tr><td class="tbl_title">Note</td><td>This selects the specified number of best hits for a given region of the
query for further evaluation. Off by default, 100 recommended if used. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.27	</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-P </td></tr>
<tr><td class="tbl_title">Function</td><td>Use multiple hit </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>Integer </td></tr>
<tr><td class="tbl_title">Example</td><td>To do single hit, use: -P 1 </td></tr>
<tr><td class="tbl_title">Note</td><td>Zero is for multiple hit, 1 for single hit. Not applicable to blastn. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.28</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-Y </td></tr>
<tr><td class="tbl_title">Function</td><td>Effective length of the search space </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Real] </td></tr>
<tr><td class="tbl_title">Example</td><td>To set this to 10000000, use: -Y 10000000 </td></tr>
<tr><td class="tbl_title">Note</td><td>This is the product of effective query length and effective database length
- actual length corrected for edge effects. Use zero for actual size. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="5" class="tbl_title">Table 4.29</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink" colspan="4">-S </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="4">Strands of the nucleotide query to use in the search </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="4">3 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="4">[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="4">To search with the reverse complement strand only, use: -S 2 </td></tr>
<tr><td class="tbl_title" rowspan="3">Note</td><td colspan="4">-S Input Code And Meaning for blastn, blastx, and tblastx. </td></tr>
<tr><td>Meaning</td><td>Input</td><td>Reverse complement</td><td>Both</td></tr>
<tr><td>Value</td><td>1</td><td>2</td><td>3</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.30</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-T </td></tr>
<tr><td class="tbl_title">Function</td><td>Produce HTML output </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">F </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[T/F] </td></tr>
<tr><td class="tbl_title">Example</td><td>To generate HTML formatted output, use: -T T</td></tr>
<tr><td class="tbl_title">Note</td><td>With -T T, if the database is from NCBI, BLAST will hot link matched
subject sequences to their actual entries in Entrez. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.31</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-u </td></tr>
<tr><td class="tbl_title">Function</td><td>Restrict search of database to the subset satisfying the query </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">N/A </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Entrez Term] in quotes </td></tr>
<tr><td class="tbl_title">Example</td><td>To restrict entries to mRNA use: -u "biomol_mrna[prop]"</td></tr>
<tr><td class="tbl_title">Note</td><td>Argument is a set of Entrez query terms. BLAST server will use the terms
to retrieve a list of GI numbers and restrict the BLAST search to entries specified by the list. Make sure valid terms are used. For example, it does
not make sense to restrict a search to genomic sequences while searching against the est database. For details, see <a href="http://www.ncbi.nlm.nih.gov/entrez/query/static/help/helpdoc.html">Entrez Help</a> </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.32</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-U </td></tr>
<tr><td class="tbl_title">Function</td><td>Use lower case filtering of FASTA sequence</td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">F </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[T/F] </td></tr>
<tr><td class="tbl_title">Example</td><td>To turn lowercase filter on, use: -U T </td></tr>
<tr><td class="tbl_title">Note</td><td>Make sure that the query sequences are in UPPERCASE and only the
filtered portions are in lowercase. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="5" class="tbl_title">Table 4.33</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink" colspan="4">-y </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="4">X dropoff value for ungapped extensions (in bits) </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="4">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="4">[Real] </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="4">To increase the dropoff to 25, use: -y 25 </td></tr>
<tr><td class="tbl_title" rowspan="3">Note</td><td colspan="4">Default setting for ungapped alignment X dropoff  (-y, in bits) </td></tr>
<tr><td>Program</td><td>blastn</td><td>megablast</td><td>others</td></tr>
<tr><td>Value</td><td>20</td><td>10</td><td>7</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="6" class="tbl_title">Table 4.34</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink" colspan="5">-Z </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="5">X dropoff value for final gapped alignment (in bits) </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="5">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="5">[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="5">To increase this dropoff to 60, use: -Z 60  </td></tr>
<tr><td class="tbl_title" rowspan="3">Note</td><td colspan="5">Large dropoff value settings may help generate longer alignment.
Default setting for ungapped alignment X dropoff  (-Z, in bits)<br/></td></tr>
<tr><td>Program</td><td>blastn</td><td>megablast</td><td>tblastx</td><td>all others</td></tr>
<tr><td>Value</td><td>50</td><td>50</td><td>25</td><td>0</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.35</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-R </td></tr>
<tr><td class="tbl_title">Function</td><td>Run rpsblast search </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">F </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[T/F] </td></tr>
<tr><td class="tbl_title">Example</td><td>To run rpsblast search, use: -R T </td></tr>
<tr><td class="tbl_title">Note</td><td>Performs rpsblast search against CDD database. Requires an
appropriate -d input. See "Remote Accessible BLAST Databases" for more information. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.36</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-n </td></tr>
<tr><td class="tbl_title">Function</td><td>Enable megablast search </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">F </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[T/F] </td></tr>
<tr><td class="tbl_title">Example</td><td>To enable megablast search, use -n T </td></tr>
<tr><td class="tbl_title">Note</td><td>Invokes megablast algorithm when set to T. -W will default to 28 and
queries will be concatenated. This will help speed up the search at the
expense of search sensitivities. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.37</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-L</td></tr>
<tr><td class="tbl_title">Function</td><td>Location on query sequence</td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">N/A</td></tr>
<tr><td class="tbl_title">Input Format</td><td>[String]    </td></tr>
<tr><td class="tbl_title">Example</td><td>To search with 100 to 400 of a query, use: -L "100,400"</td></tr>
<tr><td class="tbl_title">Note</td><td>In -L "100,400", 100 is the start and 400 the end.  </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="5" class="tbl_title">Table 4.38</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink" colspan="4">-A </td></tr>
<tr><td class="tbl_title">Function</td><td colspan="4">Multiple hits window size </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink" colspan="4">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td colspan="4">[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td colspan="4">To increase the window size to 50, use: -A 50 </td></tr>
<tr><td class="tbl_title" rowspan="3">Note</td><td colspan="4">Default -A setting for different programs </td></tr>
<tr><td>Program</td><td>blastn</td><td>megablast</td><td>all others</td></tr>
<tr><td>Value</td><td>0</td><td>0</td><td>40</td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.39</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-w </td></tr>
<tr><td class="tbl_title">Function</td><td>Frame shift penalty </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 (no penalty) </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To set OOF penalty to 10, use: -w 10 </td></tr>
<tr><td class="tbl_title">Note</td><td>Non-zero invokes OOF (Out Of Frame) algorithm for blastx. </td></tr>
</table>
<p/>
<table width="600" border="1" class="text">
<tr><td colspan="2" class="tbl_title">Table 4.40</td></tr>
<tr><td class="tbl_title" width="100">Option</td><td class="pink">-t </td></tr>
<tr><td class="tbl_title">Function</td><td>Length of the largest intron allowed in tblastn for linking HSPs </td></tr>
<tr><td class="tbl_title">Default</td><td class="pink">0 </td></tr>
<tr><td class="tbl_title">Input Format</td><td>[Integer] </td></tr>
<tr><td class="tbl_title">Example</td><td>To allow linking of HSPs 10000 letter apart, use: -t 10000 </td></tr>
<tr><td class="tbl_title">Note</td><td>Zero disables linking. Otherwise, the value specified will be used. </td></tr>
</table>
<p/>
</td></tr>

<tr><td class="tbl_title">
	<a name="5">5.  Practical usage examples</a></td></tr>
<tr><td>
Before we get into the actual use, we need to discuss the format of the input query. The only
query format blastcl3 recognizes is FASTA. In this format, the query begins with a "greater than"
sign (&gt;) initialed definition line, or defline as it is commonly known. This defline contains a 
basic description of the sequence, such as its source, the gene it represents, or ways to identify 
the sequence. It is terminated by a hard return. Actual sequence immediately follows the defline in
one or more lines each terminated by a hard return. Multiple query sequences should be
concatenated one after another. Sample query sequences are presented below for your reference.
<p/>
<pre>
&gt;gi|4557757|ref|NP_000240.1| MutL protein homolog 1
MSFVAGVIRRLDETVVNRIAAGEVIQRPANAIKEMIENCLDAKSTSIQVIVKEGGLKLIQIQDNGTGIRK
EDLDIVCERFTTSKLQSFEDLASISTYGFRGEALASISHVAHVTITTKTADGKCAYRASYSDGKLKAPPK
PCAGNQGTQITVEDLFYNIATRRKALKNPSEEYGKILEVVGRYSVHNAGISFSVKKQGETVADVRTLPNA
STVDNIRSIFGNAVSRELIEIGCEDKTLAFKMNGYISNANYSVKKCIFLLFINHRLVESTSLRKAIETVY
AAYLPKNTHPFLYLSLEISPQNVDVNVHPTKHEVHFLHEESILERVQQHIESKLLGSNSSRMYFTQTLLP
GLAGPSGEMVKSTTSLTSSSTSGSSDKVYAHQMVRTDSREQKLDAFLQPLSKPLSSQPQAIVTEDKTDIS
SGRARQQDEEMLELPAPAEVAAKNQSLEGDTTKGTSEMSEKRGPTSSNPRKRHREDSDVEMVEDDSRKEM
TAACTPRRRIINLTSVLSLQEEINEQGHEVLREMLHNHSFVGCVNPQWALAQHQTKLYLLNTTKLSEELF
YQILIYDFANFGVLRLSEPAPLFDLAMLALDSPESGWTEEDGPKEGLAEYIVEFLKKKAEMLADYFSLEI
DEEGNLIGLPLLIDNYVPPLEGLPIFILRLATEVNWDEEKECFESLSKECAMFYSIRKQYISEESTLSGQ
QSEVPGSIPNSWKWTVEHIVYKALRSHILPPKHFTEDGNILQLANLPDLYKVFERC
&gt;gi|68348711|ref|NP_001234.2| tumor necrosis factor receptor 8
MRVLLAALGLLFLGALRAFPQDRPFEDTCHGNPSHYYDKAVRRCCYRCPMGLFPTQQCPQRPTDCRKQCE
PDYYLDEADRCTACVTCSRDDLVEKTPCAWNSSRVCECRPGMFCSTSAVNSCARCFFHSVCPAGMIVKFP
GTAQKNTVCEPASPGVSPACASPENCKEPSSGTIPQAKPTPVSPATSSASTMPVRGGTRLAQEAASKLTR
APDSPSSVGRPSSDPGLSPTQPCPEGSGDCRKQCEPDYYLDEAGRCTACVSCSRDDLVEKTPCAWNSSRT
CECRPGMICATSATNSRARCVPYPICAAETVTKPQDMAEKDTTFEAPPLGTQPDCNPTPENGEAPASTSP
TQSLLVDSQASKTLPIPTSAPVALSSTGKPVLDAGPVLFWVILVLVVVVGSSAFLLCHRRACRKRIRQKL
HLCYPVQTSQPKLELVDSRPRRSSTQLRSGASVTEPVAEERGLMSQPLMETCHSVGAAYLESLPLQDASP
AGGPSSPRDLPEPRVSTEHTNNKIEKIYIMKADTVIVGTVKAELPEGRGLAGPAEPELEEELEADHTPHY
PEQETEPPLGSCSDVMLSVEEEGKEDPLPTAASGK</pre>
<p/>
Note that the file containing the query sequences has to be saved as a plain text file.
<p/>
The program runs under a command or terminal window. On PC the command window can be launched 
using "Start &#9658; Program &#9658; Accessories &#9658; Command Prompt". On Mac, the 
Terminal program icon usually is under the Utilities folder. Double click the grey icon will 
launch it.
<p/>
In the terminal window, cd to the directory containing the blastcl3, then 
run the program from there. Type "blastcl3 -" without quotes and hit return should display the 
command line options on the screen. On Mac and Unix/Linux platform, type "./blastcl3 -" without quotes.
<p/>
Since the list of available database has increased significantly, they are removed from this file.
We will document this in a separate file at a later time.
<p/>

</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;
	<a name="5.1">5.1 General nucleotide searches</a></td></tr>
<tr><td>
The primary use of nucleotide BLAST search is to identify the input query by finding if exact
match(es) are present in the database. This type of search also is used to identify the genomic
counterpart of an input mRNA sequence or vice versa. Sometimes it is also used to search with
primer pairs to identify the annealing target and possible secondary annealing sites of the primers.
<p/>
For sequences from well studied model organisms, a good approach is to search against the
refseq_rna or refseq_genomic database with Entrez limit. Alternatively, search against nr with or
without limit to the target organism can also offer good lead.
<p/>
The following example command lines search the input query file new_seq.txt against either the
refseq_rna or nr database and save the result in n_refm.out and n_nr.out, respectively.
<blockquote>
blastcl3 -p blastn -i n_seq -p blastn -d refseq_rna -o n_refm.out<br/>
blastcl3 -p blastn -i n_seq -p blastn -d nr -o n_nr.out</blockquote>

We can further restrict the search to the mouse entries in those two databases by using entrez
limit and to speed up the search by invoking the megablast algorithm. The following two options will 
accomplish that:
<blockquote>-u "mouse[organism]" -n T</blockquote>

For easy parsing of the search result, we can request that the result be returned in either XML or
"Hit Table" (tabular) format using "-m 7" or "-m 9" without quotes in the command line.
<p/>
Seaching a genomic DNA against nucleotide database, we should invoke the repeat filter to mask
the repeat region and prevent BLAST program from being inundated by spurious hits to those
regions. For human, this can be invoked by adding the following filter option to the command
line:
<blockquote>-F "m L; R"</blockquote>

Rodent specific repeat filter requires different filter call:
<blockquote>-F "R -d rodent.lib"</blockquote>
<p/>
Combining these together, the following command line searches the n_seq input nucleotide query
file against the human subset in the refseq_genomic database with low complexity and human
repeat filter and megablast algorithm.  The expect value cutoff is set to 2x10<sup>-10</sup> and 
the output is saved in refg.output:
<blockquote>blastcl3 -i n_seq -p blastn -d refseq_genomic -u "human[orgn]" -n T -F "m L; R" -e 2e-10 -o refg.output
</blockquote>

</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;<a name="5.2">5.2 General protein searches</a></td></tr>
<tr><td>
A protein BLAST search can be used to identify the input query protein or its function through 
matching to other known proteins and their annotation. One such database is refseq_protein.
The following command line searches protein sequences in my_query against this database 
using blastp. The result is saved in my_output.

<blockquote class="fixed">blatcl3 -p blastp -i my_query -d refseq_protein -o my_output</blockquote>

<p/>
For functional analysis, direct search against cdd database is more informative. Matches from
cdd search will identify the conserved functional domain(s) present in the query. Defline and
annotation from these matched domains will provide a better revelation on the function of the
query. The following command line does such a search against the cdd database (-d cdd) using
rpsblast (-R T):

<blockquote>blatcl3 -p blastp -R T -i my_query -d cdd -o my_output</blockquote>

<p/>
Specific search against pdb database can be used to identify existing structures with matching 
sequences useful for structure modeling purposes. We do not support PSI-BLAST or PHI-BLAST searches 
through blastcl3.
</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;
	<a name="5.3">5.3 Translated BLAST searches</a></td></tr>

<tr><td>
Translated searches can be very informative in revealing the possible function of the query since
the search and alignment is performed at the protein level, which is more sensitive and biologically 
relevant.
<p/>
</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<a name="5.3.1">5.3.1. blastx</a></td></tr>

<tr><td>
This program searches a nucleotide query against a protein database. It first translates the query
in all six frames and then searches those protein translations against the specified protein
database. It is useful in identify the potential protein product(s) the query may encode and may
even be able to provide information on the functions of the protein(s) should a good match to a
well characterized protein can be found.
<p/>
In the example command line below, we are searching the nucleotide sequences in my_query
against refseq_protein. The results are saved in my_oputput file.
<p/>
<blockquote>blatcl3 -p blastx -i my_query -d refseq_protein -o my_output</blockquote>
<p/>
</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<a name="5.3.2">5.3.2 tblastn</a></td></tr>

<tr><td>
This program function searches an input query protein sequence against a target nucleotide
database to find other potential protein sequences that might be encoded by those nucleotide
sequences. It is a good way to find out yet unidentified homolog/paralog of a give protein query.
During the search, the nucleotide database entries are first translated in all six frames. The query
protein is then compared against those potential products to identify the matches.
<p/>
Example given below searches the input protein query file my_query against est_human
database to try to identify human est entries that may encode proteins similar to the query. The
result is saved to my_output:
<blockquote>blastcl3 -p tblastn -i my_query -d est_human -o my_output</blockquote>
<p/>
</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<a name="5.3.3">5.3.3. tblastx</a></td></tr>

<tr><td>
<p/>
<p/>
This program function compares all six-frame translations of an input query nucleotide against
those from a nucleotide database. Since this search is very computationally expensive, we
strongly recommend that you use it with caution, employ an higher search stringency, and limit
the search to a smaller more specific subset of the database using entrez limit.
<p/>
The following command line searches the my_query against the human genomic entries in nt
database. The result is saved in my_output.

<blockquote>blastcl3 -p tblastx -i my_query -d nt -u "human[orgn] AND biomol_genomic[prop]" -o my_output
</blockquote>

<p/>
Due to the heavy computation intensity, we also recommend that users set up local standalone
blast to performing such searches if the search volume is large and/or the need is regular.
<p/>
</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;
	<a name="5.4">5.4 Genome BLAST searches</a></td></tr>

<tr><td>
<p/>
<p/>
Genome BLAST pages collect the genomic sequences and other sequences specific to an
organism in one place for easy access. In addition, the matches from searching these databases
often contain links to the graphic display on the Genome Mapviewer for that organims. Those
organism specific genomic and other sequence databases are also available for search using
blastcl3 with one major difference - there will be no link to the Map Viewer.
<p/>
</td></tr>
<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
	<a name="5.4.1">5.4.1 Microbial Genomes and Other Eukaryote Genomes</a></td></tr>

<tr><td>
<p/>
<p/>
Depending on the status of the genome, they can be finished with accompying protein data, wgs
with accompanying protein data, or wgs without accompanying protein data. The database naming
convention is "Microbial/Taxid".  The example command line below searches the protein
database for E.coli K-12 strain:
<blockquote>blastcl3 -p blastp -i my_query -d Microbial/83333 -o my_output</blockquote>

NCBI is terminating the support for BLAST searching unfinished microbial genomic sequences through the microbial genome
blast page. The recommended way is to blast against the wgs database since most of their genomic sequences are submitted
to NCBI as wgs entries. The following command line example searches the wgs entries of Bacillus anthracis: 

<blockquote class="fixed">blastcl3 -p blastn -d wgs -i my_query -o my_output -u "bacillus anthracis[orgn] AND wgs[prop]"
</blockquote>
<p/>
</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a name="5.4.2">5.4.2 Higher Genomes</a></td></tr>

<tr><td>
<p/>
<p/>
Higher genomes related databases are grouped according to orgamisms, each group has it own
unique database prefix. The genome assemblies are build-specific and they are updated when
now assemblies are made available. For example, the human genome database and other
human specific databases have the "hs_genome/" prefix. The following example command line
searches against all the available human genome assemblies:

<blockquote class="fixed">blastcl3 -p blastp -i my_query -d hs_genome/all_contig -o my_output
</blockquote>

The default filter in the human genome blast page is "low complexity, human repeat, and masking lookup table only".
To emmulate this, we can add -F "mL;R" into the command line.
<p/>
<p/>
</td></tr>

<tr><td class="tbl_title"><a name="6">6. Trouble shooting and technical assistance</a></td></tr>

<tr><td>&nbsp;</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;<a name="6.1">6.1 Errors and warnings</a></td></tr>

<tr><td>
<p/>
<p/>
Problems encountered while using blastcl3 can be caused by firewall configuration, internet
connection interruption, or NCBI server glitches, with the firewall configuration as the most
common cause. A representative error message may contain "[CONN_Open]  Cannot open
connection", "&lt;&lt;&lt; Re-establishing NETBLAST Service &gt;&gt;&gt;", or something in that order.
<p/>
Adding the following two lines in the .ncbirc file will increase the timeout setting and generate
more informative messages that are useful in debugging the problem:
<blockquote class="fixed">
TIMEOUT=300<br/>
DEBUG_PRINTOUT=DATA<br/>
</blockquote>

Search related errors from NCBI BLAST server typically are accompanied by
RID for that search. Those RIDs should be kept and sent to NCBI blast-help for trouble-shooting.
<p/>
<p/>
</td></tr>

<tr><td class="tbl_title">&nbsp;&nbsp;&nbsp;&nbsp;<a name="6">6.2 Technical assistance</a></td></tr>

<tr><td>
<p/>
<p/>
If you encounter netblast problems, please report them to blast-help alias below. We recommend
that you copy the error/warning messages displayed on the screen and provide detailed
command line, and other relevant information. Questions or comments on this document and on
BLAST in general should also be sent to blast-help alias.

<blockquote class="fixed">
<a href="mailto:blast-help@ncbi.nlm.nih.gov">
blast-help@ncbi.nlm.nih.gov</a>
</blockquote>
Questions on other NCBI resources should be sent to:
<blockquote class="fixed">
<a href="mailto:info@ncbi.nlm.nih.gov">info@ncbi.nlm.nih.gov</a>
</blockquote>
</td></tr>
<tr>
<td align="right" class="medium2"><script type="text/javascript">
//<![CDATA[
document.write("Updated on "); document.write(document.lastModified);
//]]>
</script></td>
</tr>
</tbody>
</table>
</body>
</html>
